integration-tests/deployment/keystone/deploy.go

```diff
@@ -0,0 +1,807 @@
+}
+type registerNodesResponse struct {
+	nodeIDToParams map[string]capabilities_registry.CapabilitiesRegistryNodeParams
+}
+
+func registerNodes(lggr logger.Logger, req *registerNodesRequest) (*registerNodesResponse, error) {
+	nopToNodeIDs := make(map[capabilities_registry.CapabilitiesRegistryNodeOperator][]string)
+	for nodeID, nop := range req.nodeIdToNop {
+		if _, ok := nopToNodeIDs[nop]; !ok {
+			nopToNodeIDs[nop] = make([]string, 0)
+		}
+		nopToNodeIDs[nop] = append(nopToNodeIDs[nop], nodeID)
+	}
+	nodeToRegisterNop := make(map[string]*capabilities_registry.CapabilitiesRegistryNodeOperatorAdded)
+	for _, nop := range req.nops {
+		n := capabilities_registry.CapabilitiesRegistryNodeOperator{
+			Name:  nop.Name,
+			Admin: nop.Admin,
+		}
+		nodeIDs := nopToNodeIDs[n]
+		for _, nodeID := range nodeIDs {
+			_, exists := nodeToRegisterNop[nodeID]
+			if !exists {
+				nodeToRegisterNop[nodeID] = nop
+			}
+		}
+	}
+
+	nodeIDToParams := make(map[string]capabilities_registry.CapabilitiesRegistryNodeParams)
+	for don, ocr2nodes := range req.donToOcr2Nodes {
+		caps, ok := req.donToCapabilities[don]
+		if !ok {
+			return nil, fmt.Errorf("capabilities not found for node operator %s", don)
+		}
+		var hashedCapabilityIds [][32]byte
+		for _, cap := range caps {
+			hashedCapabilityIds = append(hashedCapabilityIds, cap.id)
+		}
+		lggr.Debugw("hashed capability ids", "don", don, "ids", hashedCapabilityIds)
+
+		for _, n := range ocr2nodes {
+			if n.IsBoostrap { // bootstraps are part of the DON but don't host capabilities
+				continue
+			}
+			nop, ok := nodeToRegisterNop[n.ID]
+			if !ok {
+				return nil, fmt.Errorf("node operator not found for node %s", n.ID)
+			}
+			params, ok := nodeIDToParams[n.ID]
+
+			if !ok {
+				params = capabilities_registry.CapabilitiesRegistryNodeParams{
+					NodeOperatorId:      nop.NodeOperatorId,
+					Signer:              n.Signer,
+					P2pId:               n.P2PKey,
+					HashedCapabilityIds: hashedCapabilityIds,
+				}
+			} else {
+				// when we have a node operator, we need to dedup capabilities against the existing ones
+				var newCapIds [][32]byte
+				for _, proposedCapId := range hashedCapabilityIds {
+					shouldAdd := true
+					for _, existingCapId := range params.HashedCapabilityIds {
+						if existingCapId == proposedCapId {
+							shouldAdd = false
+							break
+						}
+					}
+					if shouldAdd {
+						newCapIds = append(newCapIds, proposedCapId)
+					}
+				}
+				params.HashedCapabilityIds = append(params.HashedCapabilityIds, newCapIds...)
+			}
+			nodeIDToParams[n.ID] = params
+		}
+	}
+
+	var uniqueNodeParams []capabilities_registry.CapabilitiesRegistryNodeParams
+	for _, v := range nodeIDToParams {
+		uniqueNodeParams = append(uniqueNodeParams, v)
+	}
+	lggr.Debugw("unique node params to add", "count", len(uniqueNodeParams))
+	tx, err := req.registry.AddNodes(req.chain.DeployerKey, uniqueNodeParams)
```

Comment:
I think you'll want to separate calldata generation from the actual call here, assuming you want the registry to be owned by MCMS. See for example here https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/add_chain.go#L131 where BuildAddDONArgs is used for in an MCMS proposal context as well as in a non-MCMS (deployment) context https://github.com/smartcontractkit/chainlink/blob/9efabb039f78c1890a4a3749fb4986d6b7947733/integration-tests/deployment/ccip/deploy_home_chain.go#L382

Reply:
will do this in a followup.

as discussed offline, the initial deployment and configuration will be done by the deployer key. changesets after that will need to go thru MCMS and will have to proposed proposals